map_model/src/pathfind/v2.rs

```diff
@@ -3,82 +3,181 @@
 //! things here will probably move into pathfind/mod.rs.
 
 use anyhow::Result;
+use serde::{Deserialize, Serialize};
+
+use geom::Duration;
 
 use crate::pathfind::uber_turns::UberTurnV2;
-use crate::{DirectedRoadID, Map, Path, PathConstraints, PathRequest, PathStep, TurnID, UberTurn};
+use crate::{
+    DirectedRoadID, Map, MovementID, Path, PathConstraints, PathRequest, PathStep, TurnID, UberTurn,
+};
+
+/// One step along a path.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum PathStepV2 {
+    /// Original direction
+    Along(DirectedRoadID),
+    /// Opposite direction, sidewalks only
+    Contraflow(DirectedRoadID),
+    Turn(MovementID),
+}
 
-/// Transform a sequence of roads representing a path into the current lane-based path, by picking
-/// particular lanes and turns to use.
-pub fn path_v2_to_v1(
+/// A path between two endpoints for a particular mode. This representation is immutable and doesn't
+/// prescribe specific lanes and turns to follow.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PathV2 {
+    steps: Vec<PathStepV2>,
+    // TODO There will be a PathRequestV2, but I'm not sure how it'll change yet.
     req: PathRequest,
-    road_steps: Vec<DirectedRoadID>,
-    uber_turns_v2: Vec<UberTurnV2>,
-    map: &Map,
-) -> Result<Path> {
-    // This is a somewhat brute-force method: run Dijkstra's algorithm on a graph of lanes and
-    // turns, but only build the graph along the path of roads we've already found. This handles
-    // arbitrary lookahead needed, and forces use of the original start/end lanes requested.
-    //
-    // Eventually we'll directly return road-based paths. Most callers will actually just use that
-    // directly, and mainly the simulation will need to expand to specific lanes, but it'll do so
-    // dynamically/lazily to account for current traffic conditions.
-    let mut graph = petgraph::graphmap::DiGraphMap::new();
-    for pair in road_steps.windows(2) {
-        for src in pair[0].lanes(req.constraints, map) {
-            for dst in pair[1].lanes(req.constraints, map) {
-                let turn = TurnID {
-                    parent: map.get_l(src).dst_i,
-                    src,
-                    dst,
-                };
-                if map.maybe_get_t(turn).is_some() {
-                    graph.add_edge(src, dst, turn);
-                }
-            }
+    cost: Duration,
+    // TODO Temporarily we'll keep plumbing these along for path_v2_to_v1 to work, but we'll
```

Comment:
When I have a bunch of TODO's related to a particular feature, especially when I merge them incrementally, I sometimes label them distinctly so I can more easily review them when I think I'm nearing "done" with that feature.

e.g. "TODO PATHV2: Temporarily we'll ..."

Otherwise it's easier to loose them with the myriad of other TODOs.

Just a suggestion - feel free to ignore.